### PR TITLE
Export MetricVec (again)

### PR DESCRIPTION
**Pull request by beorn7**
MetricVec was already exported in early versions of this library, but
nobody really used it to implement vectors of custom Metric
implementations. Now #796 has shown up with a fairly special yet valid use case
for which I'd prefer a custom implementation of a special
"auto-sampling histogram" outside of this library. Therefore, I'd like
to reinstate support for creating vectors of custom Metric
implementations.

@MikeSpreitzer @lavalamp is that helpful? If yes, I'd be grateful for a review of this PR.

I played around for quite some while with the option of a separate
package providing the tools one would need to create vectors of custom
Metric implementations. However, with the current structure of the
prometheus/client_golang/prometheus package, this leads to a lot of
complications with circular dependencies. (The new package would need
the primitives from the prometheus package, while the existing metric
vectors like GaugeVec need to import the new vector package to not
duplicate the entire MetricVec implementation. Separating vector types from the main
prometheus package is out of the question at this point because that
would be a breaking change.)

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
I would like to try developing against this.  Is there some `go.mod` tweak I can make in https://github.com/MikeSpreitzer/kubernetes/tree/add-sampling-histogram that will get me this?

**Comment by beorn7** [comment]
If you run `go get github.com/prometheus/client_golang@d1387ee3d99a5d16c8640663b6ae63663f6ae3b0` in the root of your repo (where the `go.mod` file is), your `go.mod` file should be updated accordingly. The relevant line in the `require` section should look like `github.com/prometheus/client_golang v1.7.2-0.20200910160544-d1387ee3d99a` then.

**Review by MikeSpreitzer — APPROVED** [review]
I was able to develop a client using this, see https://github.com/MikeSpreitzer/kubernetes/blob/add-sampling-histogram-vec/staging/src/k8s.io/component-base/metrics/prometheusextension/sampling-histogram.go (based off https://github.com/kubernetes/kubernetes/pull/94672).

**Comment by beorn7** [comment]
@MikeSpreitzer Glad to see that. Your approach looks good in general. I'll comment in #796 in more detail.

I'll get a second opinion from someone of prometheus-team before I merge this. E.g. @bwplotka as a heavy user of client_golang, do you think this makes sense to add this "construction kit for metric vectors" to client_golang?

**Comment by beorn7** [comment]
@bwplotka :bellhop_bell: Could you have a look and let me know if you think this is useful to have in client_golang?

**Comment by beorn7** [comment]
@bwplotka another attempt to get your attention… :pray:

**Comment by lavalamp** [comment]
cc @logicalhan

File: prometheus/vec.go
Comment (bwplotka) [inline prometheus/vec.go:26]:
```suggestion
// CounterVec, SummaryVec, and HistogramVec. It is exported so that it can be
```
(:

File: prometheus/vec.go
Comment (bwplotka) [inline prometheus/vec.go:27]: Super useful comment. I would suggesting adding new line to separate this tip for `Vec` implementation for custom type with upper summary?

File: prometheus/vec.go
Comment (bwplotka) [inline prometheus/vec.go:127]: :+1:

**Review by bwplotka — APPROVED** [review]
Love it, just a tiny suggestion and typo to fix and LGTM :tada: 

Sorry for the lag, I am behind all GH notifications, accidentally seen this ping... 

 

Will do my best next time, if I'm not responding after 3-4d feel free to ping me on Slack (:

File: prometheus/vec.go
Comment (beorn7) [reply prometheus/vec.go:27]: Sure. Done.

File: prometheus/vec.go
Comment (beorn7) [reply prometheus/vec.go:26]: Fixed.

**Comment by beorn7** [comment]
Thanks @bwplotka . I have addressed the comments and will merge on green.

**Pull request merged**
